src/adapter/src/catalog/apply.rs

```diff
@@ -504,6 +512,236 @@ impl CatalogState {
+                            database_spec: ResolvedDatabaseSpecifier::Ambient,
+                            schema_spec: SchemaSpecifier::Id(schema_id),
+                        },
+                        item: typ.name.to_owned(),
+                    },
+                    CatalogItem::Type(Type {
+                        create_sql: None,
+                        details: typ.details.clone(),
+                        desc,
+                        resolved_ids: ResolvedIds(BTreeSet::new()),
+                    }),
+                    MZ_SYSTEM_ROLE_ID,
+                    PrivilegeMap::from_mz_acl_items(vec![
+                        rbac::default_builtin_object_privilege(mz_sql::catalog::ObjectType::Type),
+                        rbac::owner_privilege(mz_sql::catalog::ObjectType::Type, MZ_SYSTEM_ROLE_ID),
+                    ]),
+                );
+            }
+
+            Builtin::Func(func) => {
+                // This OID is never used. `func` has a `Vec` of implementations and
+                // each implementation has its own OID. Those are the OIDs that are
+                // actually used by the system.
+                let oid = INVALID_OID;
+                self.insert_item(
+                    id,
+                    oid,
+                    name.clone(),
+                    CatalogItem::Func(Func { inner: func.inner }),
+                    MZ_SYSTEM_ROLE_ID,
+                    PrivilegeMap::default(),
+                );
+            }
+
+            Builtin::Source(coll) => {
+                let mut acl_items = vec![rbac::owner_privilege(
+                    mz_sql::catalog::ObjectType::Source,
+                    MZ_SYSTEM_ROLE_ID,
+                )];
+                acl_items.extend_from_slice(&coll.access);
+
+                self.insert_item(
+                    id,
+                    coll.oid,
+                    name.clone(),
+                    CatalogItem::Source(Source {
+                        create_sql: None,
+                        data_source: DataSourceDesc::Introspection(coll.data_source),
+                        desc: coll.desc.clone(),
+                        timeline: Timeline::EpochMilliseconds,
+                        resolved_ids: ResolvedIds(BTreeSet::new()),
+                        custom_logical_compaction_window: coll.is_retained_metrics_object.then(
+                            || {
+                                self.system_config()
+                                    .metrics_retention()
+                                    .try_into()
+                                    .expect("invalid metrics retention")
+                            },
+                        ),
+                        is_retained_metrics_object: coll.is_retained_metrics_object,
+                    }),
+                    MZ_SYSTEM_ROLE_ID,
+                    PrivilegeMap::from_mz_acl_items(acl_items),
+                );
+            }
+        }
```

Comment:
This is mostly copied moved from `open.rs`, the one difference is `state.` is replaced with `self.`.